### ISSUE TITLE
Replace older `sct_testing` tests with new `pytest` tests

### ISSUE THREAD

**Issue by joshuacwnewton** [spinalcordtoolbox/spinalcordtoolbox#3093]
In #2893, we decided to move away from `sct_testing` and instead start using `pytest`. As a short-term solution, we minimally wrapped the old tests so that they can be run as-is using `pytest`. As a long-term solution, though, we should start replacing the old `sct_testing` tests with equivalent pure `pytest` tests. 

<details>
<summary> List of sct_testing files to convert </summary>

## Tests already converted to `pytest`
- [X] test_sct_deepseg_sc.py
- [X] test_sct_get_centerline.py

## Tests that only run the CLI command (PR #3244)

Listed as "Not implemented.":
- [x] test_sct_compute_mtr.py
- [x] test_sct_compute_snr.py
- [x] test_sct_concat_transfo.py
- [x] test_sct_create_mask.py
- [x] test_sct_deepseg_lesion.py
- [x] test_sct_dmri_compute_bvalue.py
- [x] test_sct_dmri_compute_dti.py
- [x] test_sct_dmri_concat_b0_and_dwi.py
- [x] test_sct_dmri_concat_bvals.py
- [x] test_sct_dmri_concat_bvecs.py
- [x] test_sct_dmri_transpose_bvecs.py
- [x] test_sct_flatten_sagittal.py
- [x] test_sct_fmri_compute_tsnr.py
- [x] test_sct_fmri_moco.py
- [x] test_sct_label_vertebrae.py
- [x] test_sct_maths.py
- [x] test_sct_merge_images.py
- [x] test_sct_process_segmentation.py
- [x] test_sct_qc.py
- [x] test_sct_smooth_spinalcord.py
- [x] test_sct_straighten_spinalcord.py
- [x] test_sct_warp_template.py

Tests that are currently broken:
- [x] test_sct_denoising_onlm.py

## Tests that run the CLI command + `test_integrity` (PR #3246)

Tests that only perform very basic `test_integrity` checks (e.g. file existence check, value comparison with cached values, etc.):
- [x] test_sct_analyze_lesion.py
- [x] test_sct_compute_ernst_angle.py
- [x] test_sct_compute_hausdorff_distance.py
- [x] test_sct_convert.py
- [x] test_sct_crop_image.py
- [x] test_sct_dice_coefficient.py
- [x] test_sct_extract_metric.py
- [x] test_sct_resample.py

Tests with slightly more complex assertions (e.g. some calculations are involved):
- [x] test_sct_analyze_texture.py
- [x] test_sct_compute_mscc.py
- [x] test_sct_deepseg_gm.py
- [x] test_sct_detect_pmj.py
- [x] test_sct_dmri_separate_b0_and_dwi.py
- [x] test_sct_propseg.py
- [x] test_sct_register_multimodal.py

Complex tests (e.g. a more involved setup, or multiple different checks all in one test, etc.):
- [x] test_sct_apply_transfo.py
- [x] test_sct_dmri_moco.py
- [x] test_sct_image.py
- [x] test_sct_register_to_template.py

Tests that are currently broken:
- [x] test_sct_label_utils.py

</details>

**Comment by Drulex** [comment]
What is the added value of replacing the functional tests if they are working correctly?

Maybe:
- If a functional test needs to be extended it could be refactored to use the new framework in order to leverage the new capabilities.
- It would be a good exercise for new contributors to get to know a certain module.

Other than that I don't see a strong motivation to do so, what do you think?

**Comment by joshuacwnewton** [comment]
> What is the added value of replacing the functional tests if they are working correctly?

Truthfully, I hadn't thought too hard about this, so I'm happy you've asked this question.

To add to your list:

1. Old tests are run using subprocesses. If we rewrote the tests to instead run using a script's `main` function, I think we would see a bit of a speedup. 
    * Demo: I wrote 2 versions of [a test that calls every script](https://github.com/neuropoly/spinalcordtoolbox/commit/f94200c2c9dcfe39fbaf16c4744d2d7f32877133) with no args: one version used `subprocess` and the other used `<sct_script>.main()`. The main version reduced the test time [from ~50s to ~5s](https://github.com/neuropoly/spinalcordtoolbox/pull/3156#discussion_r556690925).
2. The existing tests aren't descriptive. Some of them just run the scripts with no validation, and others test all kinds of functionality, but each test has the same generic name. So, it takes a little extra effort to understand what each test is doing.
3. Because the tests are split between two places (`unit_testing/` for pytest and `testing/` for sct_testing), contributors have to bounce between 2 test modules per script, and switch between mental models for testing frameworks.
4. `pytest` has been battle-tested by many projects, but `sct_testing` itself is (relatively) untested, so I think it's more likely to have obscure bugs/quirks we haven't discovered yet (like the one in your edit).
5. The `sct_testing` framework is, for me, hard to understand. For onboarding people to the project, it would be simpler to say "we use pytest", but there's this lurking behemoth behind the scenes that takes time/effort to wrap your head around. I find that reading through `sct_testing.py` (e.g. for debugging) is much more difficult compared to reading pytest documentation and GH/SO threads. 

So for me, it's a case of "is the upfront effort worth saving future maintenance costs," but I'm not sure atm.

**Comment by joshuacwnewton** [comment]
This has been in the back of my head these past few months, and I ran into a few more reasons:

7. **Debugging:** Because `sct_testing` uses subprocesses, it's less straightforward to attach a debugger. Especially considering a lot of the lab uses PyCharm (myself included), their various debugging interfaces don't play nicely with `sct_testing`.
8. **Coverage:** As part of #2702, if/when we introduce coverage checking, AFAIK subprocesses [are trickier to work with](https://coverage.readthedocs.io/en/latest/subprocess.html).

**Comment by joshuacwnewton** [comment]
>  4. **Test organization:** Because the tests are split between two places (`unit_testing/` for pytest and `testing/` for sct_testing), contributors have to bounce between 2 test modules per script (and switch between mental models for testing frameworks) to make sure they're not duplicating anything.

To expand on this point specifically, I made a separate issue: https://github.com/neuropoly/spinalcordtoolbox/issues/3219

**Comment by joshuacwnewton** [comment]
**Caveat:** We often still recommend the `sct_testing` command as a way for users to verify their installation. If we move away from `sct_testing`, then we would likely begin recommending `pytest` for users to verify their installation instead.

However, then #2911 would become a nuisance for our users. (By comparison, `sct_testing` runs everything within a temporary directory.)

So, I may have to pump the breaks on this idea until #2911 is addressed.

**Comment by jcohenadad** [comment]
indeed, `sct_testing` is currently *super useful* for us

**Comment by Drulex** [comment]
> However, then #2911 would become a nuisance for our users. (By comparison, `sct_testing` runs everything within a temporary directory.)

I think that the test artifacts are created in the working directory so it may be a matter of simply running the tests from a tmpfs.

**Comment by joshuacwnewton** [comment]
> I think that the test artifacts are created in the working directory so it may be a matter of simply running the tests from a tmpfs (e.g: cd /tmp && pytest $SCT_DIR or something).

This is true. My hope, though, would be to make sure the tests clean up after themselves regardless of the working directory. 

Thankfully, there are only 8 tests that put files in the working directory, so it shouldn't be too much work for me to add a more long-term fix. I made a comment with more details over in https://github.com/neuropoly/spinalcordtoolbox/issues/2911#issuecomment-779482628 (which is probably a better place for discussion about this).

**Comment by joshuacwnewton** [comment]
I've written a quick overview of the existing `sct_testing` tests in the issue description: https://github.com/neuropoly/spinalcordtoolbox/issues/3093#issue-764595139.

Surprisingly, **more than half** of the tests (26/45) aren't doing much right now (reasons: broken, or checks not implemented, or they have already been converted). 

This is a bit concerning, but at least it means it will be easier to migrate the tests than I first thought. :sweat_smile:

**Comment by joshuacwnewton** [comment]
> indeed, `sct_testing` is currently _super useful_ for us

I wanted to address this concern: It is possible to [`@pytest.mark`](https://docs.pytest.org/en/stable/example/markers.html#mark-examples) certain tests with metadata. For example:

```python
@pytest.mark.sct_testing
def test_sct_compute_mtr_no_checks():
    """Run the CLI script without checking results."""
    sct_compute_mtr.main(argv=['-mt0', 'mt/mt0.nii.gz', '-mt1', 'mt/mt1.nii.gz'])
```

Then, if the command `pytest -m sct_testing` is run, only the tests marked with `sct_testing` data will be run, which would preserve the functionality of the `sct_testing` command.

**Comment by joshuacwnewton** [comment]
In the process of converting the tests, I found 2 tests that were silently broken, but we did not know because `sct_testing` did not alert us.

<details>
<summary> <code>test_cli_sct_resample</code> </summary>

This test makes an assertion about expected output dimensions:

```python
    if not np.all(np.round(i, 4) == np.round(j, 4) for i, j in zip(dims, param_test.results_dims[index_args])):
        param_test.output += 'WARNING: dimensions and spacing different from expected.' \
```

This assertion doesn't actually fail when it should, though. When I rewrote the assertion, I got a failing test.

```python
    for actual_val, expected_val in zip(actual_dim, expected_dim):
        assert actual_val == pytest.approx(expected_val)
```

</details>

<details>
<summary> <code>test_cli_sct_register_multimodal</code> </summary>
 
https://github.com/neuropoly/spinalcordtoolbox/blob/7dc458c7f827a4ee1453fb3f97a7b4f70551f5c3/testing/test_sct_register_multimodal.py#L41-L46

From this snippet, the integrity testing only runs if `fname_gt` is present.

But that file is never actually defined anywhere, so the integrity check gets skipped every time. 

`sct_testing` even hides this fact! No verbose output (looks OK):

```console
> sct_testing --function sct_register_multimodal

--
Spinal Cord Toolbox (git-jn/3093-migrate-simple-sct_testing-tests-d24f931b16d3646e73d95ade54f29eabbfa75ab6*)

sct_testing --function sct_register_multimodal
--

Creating temporary folder (/tmp/sct-20210224093900.725928-m672e3gk)
Will run through the following tests:
- in parallel with 1 jobs: sct_register_multimodal
Checking sct_register_multimodal....................[OK]
status: [0]
Finished! Elapsed time: 13s
```

With `--verbose` output, integrity testing doesn't happen, only outputting  `"--> N/A"`:

```console
> sct_testing --function sct_register_multimodal --verbose

--
Spinal Cord Toolbox (git-jn/3093-migrate-simple-sct_testing-tests-d24f931b16d3646e73d95ade54f29eabbfa75ab6*)

sct_testing --function sct_register_multimodal --verbose
--
   ====================================================================================================
   sct_register_multimodal -i mt/mt0.nii.gz -iseg mt/mt0_seg.nii.gz -d mt/mt1.nii.gz -dseg mt/mt1_seg.nii.gz -o mt0_reg.nii.gz -param step=1,algo=columnwise,type=seg,smooth=1 -x linear -r 0
   ====================================================================================================
   
   
   --
   Spinal Cord Toolbox (git-jn/3093-migrate-simple-sct_testing-tests-d24f931b16d3646e73d95ade54f29eabbfa75ab6*)
   
   sct_register_multimodal -i mt/mt0.nii.gz -iseg mt/mt0_seg.nii.gz -d mt/mt1.nii.gz -dseg mt/mt1_seg.nii.gz -o mt0_reg.nii.gz -param step=1,algo=columnwise,type=seg,smooth=1 -x linear -r 0
   --
   
   # stdout hidden for brevity
   
   Finished! Elapsed time: 0s
   
   ====================================================================================================
   INTEGRITY TESTING
   ====================================================================================================
   
   
   --> N/A
status: [0]
Finished! Elapsed time: 13s

```

</details>

For the second test, that kind of hidden quirk is part of why I think it's important we switch to `pytest`. :sweat_smile:

**Comment by joshuacwnewton** [comment]
Just to make the status of this issue clear: The work is all completed... the trouble is coordinating reviews for the work, considering there would be at minimum ~90 changed files. (45 tests in `testing/` would be deleted, and 45 tests would be added in `unit_testing/`). 

I could split this up into several PRs, but then we would have a partially-functioning `sct_testing`... so I would prefer to start this right after the next release (so it can be completed before the next-next-release).

Despite the workload, I still believe this would be quite valuable for making it easier for external contributors to write tests, see for example https://github.com/neuropoly/spinalcordtoolbox/pull/3320#pullrequestreview-634912606.
